### PR TITLE
[routing-manager] prepare route/prefix info options in place

### PR DESCRIPTION
**Pull request by abtink**
This commit contains smaller enhancements in `RoutingManager `
and `Prefix/RouteInfoOption`.

This commit changes how the options are added in a message so that we
directly construct the option in the message buffer (instead of
creating it separately and then copying the bytes).

It also changes `RouteInfoOption::IsValid()` to validate that the
option length can fit the specified prefix length.

### CODE REVIEW

File: src/core/border_router/router_advertisement.cpp [hunk 1]
```diff
@@ -113,62 +86,85 @@ void PrefixInfoOption::SetPrefix(const Ip6::Prefix &aPrefix)
...
+    // - 3 otherwise.
+
+    if (aPrefixLength == kMaxPrefixLenForOptionLen1)
+    {
+        length = 1;
+    }
+    else if (aPrefixLength <= kMaxPrefixLenForOptionLen2)
+    {
+        length = 2;
+    }
```
Comment (wgtdkp) [inline src/core/border_router/router_advertisement.cpp:160]: nit: just for aligning with the comment :)
```suggestion
    uint8_t length;

    // The Option Length can be 1, 2, or 3 depending on the prefix
    // length
    //
    // - 1 when prefix len is zero.
    // - 2 when prefix len is less then or equal to 64.
    // - 3 otherwise.

    if (aPrefixLength == kMaxPrefixLenForOptionLen1)
    {
        length = 1;
    }
    else if (aPrefixLength <= kMaxPrefixLenForOptionLen2)
    {
        length = 2;
    }
    else
    {
        length = 3;
    }
```

File: src/core/border_router/router_advertisement.hpp [hunk 2]
```diff
@@ -255,31 +257,52 @@ class PrefixInfoOption : public Option
...
+    //  |                           Reserved2                           |
+    //  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    //  |                                                               |
+    //  +                                                               +
+    //  |                                                               |
+    //  +                            Prefix                             +
+    //  |                                                               |
+    //  +                                                               +
+    //  |                                                               |
+    //  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
Comment (wgtdkp) [inline src/core/border_router/router_advertisement.hpp:299]: nice layout diagram!

**Review by wgtdkp — APPROVED** [review]
LGTM 👍

File: src/core/border_router/router_advertisement.hpp [hunk 3]
```diff
@@ -150,80 +143,89 @@ class Option
     static const Option *GetNextOption(const Option *aCurOption, const uint8_t *aBuffer, uint16_t aBufferLength);
 
     /**
-     * This method tells whether this option is valid.
+     * This method indicates whether or not this option is valid.
      *
-     * @return  A boolean that indicates whether this option is valid.
+     * @retval The option is valid.
+     * @retval The option is not valid.
```
Comment (jwhui) [inline src/core/border_router/router_advertisement.hpp:149]:
```suggestion
     * @retval  TRUE   The option is valid.
     * @retval  FALSE  The option is not valid.
```

File: tests/toranj/openthread-core-toranj-config.h [hunk 4]
```diff
@@ -67,6 +67,14 @@
  */
 #define OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE 1
 
+/**
+ * @def OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+ *
+ * Define to 1 to enable Border Routing support.
+ *
+ */
+#define OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE 1
```
Comment (jwhui) [inline tests/toranj/openthread-core-toranj-config.h:76]: Is this change related to this PR?

Comment (abtink) [reply tests/toranj/openthread-core-toranj-config.h:76]: No. Not directly related, but enabling it under `toranj` does help cover the `border_router` src codes under more CI builds and should be harmless since on `toranj` tests it is not enabled.

**Review by jwhui — APPROVED** [review]
